# Scanpy Use Cases and Tutorials

Start off with one of the use cases below. The following template might be helpful: [myexample_template.ipynb](myexample_template.ipynb). Examples for using the command line are [17-05-22](https://github.com/theislab/scanpy_usage/tree/master/170522_visualizing_one_million_cells) and [17-05-01](https://github.com/theislab/scanpy_usage/tree/master/170501_moignard15).

* [17-05-29](https://github.com/theislab/scanpy_usage/tree/master/170529_images) | We discuss the pseudotime-based reconstruction of cell cycle from image data and compare it with the deep-learning based approach of [Eulenberg, KÃ¶hler *et al.*, Nat. Comms. (2017)](https://doi.org/10.1101/081364).

* [17-05-22](https://github.com/theislab/scanpy_usage/tree/master/170522_visualizing_one_million_cells) | Visualizing the [publicly available](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1M_neurons) 1.3 mio brain cell dataset from 10X Genomics using the command line.

* [17-05-05](https://github.com/theislab/scanpy_usage/tree/master/170505_seurat) | We reproduce most of the [Guided Clustering Tutorial](http://satijalab.org/seurat/pbmc-tutorial.html) of [*Seurat*](http://satijalab.org/seurat/) [(Macosko *et al.*, Cell 2015)](https://doi.org/10.1016/j.cell.2015.05.002).

* [17-05-03](https://github.com/theislab/scanpy_usage/tree/master/170503_zheng17) | Analyzing 68 000 cells from [(Zheng *et al.*, Nat. Comms. 2017)](https://doi.org/10.1038/ncomms14049), we find that Scanpy is about a factor 5 to 16 faster and more memory efficient than the [*Cell Ranger*](https://github.com/10XGenomics/single-cell-3prime-paper/tree/master/pbmc68k_analysis) R kit for secondary analysis.

* [17-05-02](https://github.com/theislab/scanpy_usage/tree/master/170502_haghverdi16) | We reproduce the results of the Diffusion Pseudotime (DPT) paper of [Haghverdi *et al.*, Nat. Meth. (2016)](https://doi.org/10.1038/nmeth.3971). Note that DPT has recently been very [favorably discussed](https://doi.org/10.1101/110668) by the authors of [Monocle](http://cole-trapnell-lab.github.io/monocle-release/articles/v2.0.0/).

* [17-05-02](https://github.com/theislab/scanpy_usage/tree/master/170502_paul15) | Diffusion Pseudotime analysis reconstructs myeloid and erythroid differentiation for single-cell RNAseq data of [Paul *et al.*, Cell (2015)](http://doi.org/10.1016/j.cell.2015.11.013).

* [17-05-01](https://github.com/theislab/scanpy_usage/tree/master/170501_moignard15) | Diffusion Pseudotime analysis reconstructs early hematopoiesis from  single-cell qPCR data of [Moignard *et al*, Nat. Biotechn. (2015)](https://doi.org/10.1038/nbt.3154).
